compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -743,8 +743,166 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
+                finder.visit_expr(e);
+                if finder.found {
+                    // The current scope includes the declaration of the binding we're accessing, we
+                    // can't look up any further for loops.
+                    break;
+                }
+            }
+            if parent.is_none() {
+                parent = Some(e);
+            }
+            match e.kind {
+                hir::ExprKind::Let(_) => {
+                    match tcx.parent_hir_node(e.hir_id) {
+                        hir::Node::Expr(hir::Expr {
+                            kind: hir::ExprKind::If(cond, ..), ..
+                        }) => {
+                            let mut finder = Finder { hir_id: expr.hir_id, found: false };
+                            finder.visit_expr(cond);
+                            if finder.found {
+                                // The expression where the move error happened is in a `while let`
+                                // condition Don't suggest clone as it will likely end in an
+                                // infinite loop.
+                                // while let Some(_) = foo(non_copy.clone()) { }
+                                // ---------                       ^^^^^^^^
+                                can_suggest_clone = false;
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+                hir::ExprKind::Loop(..) => {
+                    outer_most_loop = Some(e);
+                }
+                _ => {}
+            }
+        }
+
+        /// Look for `break` expressions within any arbitrary expressions. We'll do this to infer
+        /// whether this is a case where the moved value would affect the exit of a loop, making it
+        /// unsuitable for a `.clone()` suggestion.
+        struct BreakFinder {
+            found_break: bool,
+        }
+        impl<'hir> Visitor<'hir> for BreakFinder {
+            fn visit_expr(&mut self, ex: &'hir hir::Expr<'hir>) {
+                if let hir::ExprKind::Break(..) = ex.kind {
+                    self.found_break = true;
+                }
+                hir::intravisit::walk_expr(self, ex);
+            }
+        }
+        if let Some(in_loop) = outer_most_loop
+            && let Some(parent) = parent
+            && let hir::ExprKind::MethodCall(..) | hir::ExprKind::Call(..) = parent.kind
+        {
+            // FIXME: We could check that the call's *parent* takes `&mut val` to make the
+            // suggestion more targeted to the `mk_iter(val).next()` case. Maybe do that only to
+            // check for wheter to suggest `let value` or `let mut value`.
+
+            let span = in_loop.span;
+            let mut finder = BreakFinder { found_break: false };
+            finder.visit_expr(in_loop);
+            let sm = tcx.sess.source_map();
+            if (finder.found_break || true)
```

Comment:
I presume the `|| true` was for testing?
```suggestion
            if finder.found_break
```

Reply:
I see you removed it in the next commit